_posts/2016-02-29-avoid-runtime-errors-in-react-with-flow.markdown

```diff
@@ -0,0 +1,372 @@
+
+Why Flow? Our team settled on Flow because we already had multiple fairly substantial React codebases with complicated-enough build systems that we wanted to avoid modifying (the build systems), which pretty much discounted TypeScript, at least for the immediate gains we were seeking. Luckily, Flow gave us the opportunity to incrementally add types to our applications with a simple `/* @flow */` pragma and little else.
+
+If you haven't given it a chance yet I strongly advise [giving it five minutes](https://signalvnoise.com/posts/3124-give-it-five-minutes). In fact, despite balking at some of the syntax at first, the first time I saw a `Property cannot be accessed on possibly null or undefined value` on a [Maybe Type](http://flowtype.org/docs/nullable-types.html#_) I pretty much spammed our team's Slack channel in excitement.
+
+It was, as the thesis states, awesome.
+
+## Type Safety and React
+
+Unlike what some [recent rumors](https://www.reddit.com/r/reactjs/comments/44yqge/fyi_proptypes_are_on_their_way_out/czubij1) may have implied, `PropTypes` are not going anywhere, at least any time soon. However, that doesn't mean they're perfect; there are a number of places where they fall short, notably:
+
+1) `PropTypes` are not statically analyzable. When you encounter an error it's during development and only once the problem component has rendered. That's far too late, and the fact that it's caught at runtime means you're likely to miss errors that could end up biting you when you roll out to production.
+
+2) They add unneeded `noop` bytes in production. Yes, not the biggest of deals, and it [is possible to remove them](https://github.com/oliviertassinari/babel-plugin-transform-react-remove-prop-types), but that's Yet Another Plugin for your growing webpack + babel + rosetta stone build pipeline.
+
+3) As the name implies, the type checking is limited to component properties: they cannot be used to type check any other parts of your codebase.
+
+4) Finally, `PropTypes` are simply less powerful. They can, to a degree, ensure that the props entering a component match your expectations, but that's where their utility ends; they can make no guarantees that you, for example, do not try to access non-existent properties on objects passed through props, nor do `PropTypes` check _how_ those props are used throughout a component.
+
+Flow solves all of these concerns and is an invaluable addition to the typical workflow, perhaps as a [pre-commit hook](http://technologyadvice.github.io/the-power-of-the-precommit/) or as part of your CI process. But enough with the words already, let's write some code. Everyone is rightfully sick of [TodoMVC](http://todomvc.com/), so we'll change it up by doing some TodoMVC _with Flow_.
```

Comment:
Nice internal backlink! The SEO crowd is smiling. If desired, you can remove the `http://technologyadvice.github.io` from the beginning of the link to make the link root-relative.